File: src/setcontrolsmenu.py
from typing import TYPE_CHECKING
from .menu import Menu
from .buttons import TextButton
from .common import GameState
import pygame

if TYPE_CHECKING:
    from .game import Game
    from .controls import Controls

class SetControlsMenu(Menu):
    NUM_OF_CONTROLS = 11
    
    LINKED_CONTROLS = [
        -1, 2, 1, -1, -1, 6, 5, 8, 7, 10, 9
    ]
    
    CONTROL_STRINGS = [
        "Fire Weapon", "Change Weapon Up", "Change Weapon Down",
        "Use Jump Jets", "Use Shield",
        "Move Tank Left", "Move Tank Right",
        "Rotate Gun Left", "Rotate Gun Right",
        "Increase Gun Power", "Decrease Gun Power"
    ]
    
    AXIS_NAMES = [
        "Joystick/Pad Right", "Joystick/Pad Left",
        "Joystick/Pad Up", "Joystick/Pad Down",
        "Axis 3 (-)", "Axis 3 (+)",
        "Axis 4 (-)", "Axis 4 (+)"
    ]

    def __init__(self, game: 'Game', layout: int):
        super().__init__(game)
        self._layout = layout
        self._controls = game.get_controls()
        
        self._control_buttons = [None] * self.NUM_OF_CONTROLS
        self._control_key = [0] * self.NUM_OF_CONTROLS
        
        for i in range(self.NUM_OF_CONTROLS):
            self._control_buttons[i] = TextButton(self, -3.0, 5.0 - i * 0.8, 0.5, self.CONTROL_STRINGS[i])
            self._control_key[i] = self._controls.get_control(layout, i)
            
        self._reset_button = TextButton(self, 0.0, -5.0, 0.7, "Reset To Defaults")
        self._done_button = TextButton(self, 0.0, -6.0, 0.7, "Done")
        
        self._waiting_for_key = -1
        
    def update(self, time: float) -> int:
        self.update_background(time)
        
        for i in range(self.NUM_OF_CONTROLS):
             if self._control_buttons[i].update():
                 self._waiting_for_key = i
                 for j in range(self.NUM_OF_CONTROLS):
                     self._control_buttons[j].enable(False)
                     
        if self._waiting_for_key != -1:
            if self._layout < 2:
                found_key = -1
                for event in pygame.event.get():
                    if event.type == pygame.QUIT:
                        pygame.event.post(event)
                        break
                    
                    if event.type == pygame.KEYDOWN:
                        found_key = event.key
                        break
                        
                if found_key != -1:
                    self._control_key[self._waiting_for_key] = found_key
                    self._waiting_for_key = -1
                    for j in range(self.NUM_OF_CONTROLS):
                        self._control_buttons[j].enable(True)
            else:
                joy_idx = self._layout - 2
                found = False
                interface = self._game.get_interface()
                
                for b in range(16):
                    if interface.get_joystick_button(joy_idx, b):
                        if self._control_key[self._waiting_for_key] >= 100 and \
                           self.LINKED_CONTROLS[self._waiting_for_key] != -1:
                               self._control_key[self.LINKED_CONTROLS[self._waiting_for_key]] = -1
                        
                        self._control_key[self._waiting_for_key] = b
                        found = True
                        break
                
                if not found:
                    for a in range(8):
                        val = interface.get_joystick_axis(joy_idx, a)
                        if val > 0.5:
                            self._control_key[self._waiting_for_key] = 100 + (a * 2)
                            linked = self.LINKED_CONTROLS[self._waiting_for_key]
                            if linked != -1: self._control_key[linked] = 101 + (a * 2)
                            found = True
                            break
                        elif val < -0.5:
                            self._control_key[self._waiting_for_key] = 101 + (a * 2)
                            linked = self.LINKED_CONTROLS[self._waiting_for_key]
                            if linked != -1: self._control_key[linked] = 100 + (a * 2)
                            found = True
                            break
                            
                if found:
                    self._waiting_for_key = -1
                    for j in range(self.NUM_OF_CONTROLS):
                        self._control_buttons[j].enable(True)
                        
        if self._done_button.update():
            for i in range(self.NUM_OF_CONTROLS):
                self._controls.set_control(self._layout, i, self._control_key[i])
            return GameState.CONTROLLERS_MENU
            
        if self._reset_button.update():
            self._controls.reset_to_default(self._layout)
            for i in range(self.NUM_OF_CONTROLS):
                self._control_key[i] = self._controls.get_control(self._layout, i)
                
        return GameState.CURRENT_STATE

    def draw(self):
        self.draw_background()
        boxes = [
            [(-7.0, -4.0), (7.0, -4.0), (7.0, 6.0), (-7.0, 6.0)],
            [(-7.0, -6.6), (7.0, -6.6), (7.0, -4.4), (-7.0, -4.4)]
        ]
        dark = (0, 0, 0, 128)
        for b in boxes:
            self.draw_game_polygon(b, dark)
            
        brown = (153, 76, 0, 128)
        
        for i in range(self.NUM_OF_CONTROLS):
            y_top = 5.3 - i * 0.8
            y_bot = 4.6 - i * 0.8
            lb = [(-6.0, y_top), (0.0, y_top), (0.0, y_bot), (-6.0, y_bot)]
            rb = [(0.1, y_top), (6.0, y_top), (6.0, y_bot), (0.1, y_bot)]
            self.draw_game_polygon(lb, brown)
            self.draw_game_polygon(rb, brown)
            
        bb1 = [(-4.0, -5.4), (4.0, -5.4), (4.0, -4.6), (-4.0, -4.6)]
        bb2 = [(-4.0, -6.4), (4.0, -6.4), (4.0, -5.6), (-4.0, -5.6)]
        self.draw_game_polygon(bb1, brown)
        self.draw_game_polygon(bb2, brown)

        title = f"Editing Keyboard Layout {self._layout + 1}" if self._layout < 2 else f"Editing Joystick Layout {self._layout - 1}"
        self._ui.draw_centered_text(0.0, 6.5, title, style=self._ui.style(0.6, (1.0, 1.0, 1.0), shadow=True))
        
        if self._waiting_for_key != -1:
            self._ui.draw_centered_text(
                0.0,
                -4.0,
                f"Press Button for '{self.CONTROL_STRINGS[self._waiting_for_key]}'",
                style=self._ui.style(0.6, (0.5, 0.5, 0.5), shadow=True),
            )

        value_style = self._ui.style(0.5, (0.0, 1.0, 1.0), spacing=0.4)
        for i in range(self.NUM_OF_CONTROLS):
            key = self._control_key[i]
            txt = "<Undefined>"
            
            if self._layout < 2:
                if key > 0:
                    txt = pygame.key.name(key)
            else:
                if key == -1: txt = "<Undefined>"
                elif key < 100: txt = f"Joy Button {key + 1}"
                else: 
                     idx = key - 100
                     if idx < len(self.AXIS_NAMES):
                         txt = self.AXIS_NAMES[idx]
                     else:
                         txt = f"Axis {idx}"
            
            self._ui.draw_centered_text(3.0, 4.7 - i * 0.8, txt, style=value_style)
            
        for i in range(self.NUM_OF_CONTROLS):
            self._control_buttons[i].draw()
            
        self._reset_button.draw()
        self._done_button.draw()

File: groundfire-0.25/src/setcontrolsmenu.cc
////////////////////////////////////////////////////////////////////////////////
//
//               Groundfire
//
////////////////////////////////////////////////////////////////////////////////
//
// Copyright (c) 2004, Tom Russell (tom@groundfire.net)
//
// This file is part of the Groundfire project, distributed under the MIT 
// license. See the file 'COPYING', included with this distribution, for a copy
// of the full MIT licence.
//
////////////////////////////////////////////////////////////////////////////////
//
//   File name : setcontrolsmenu.cc
//
//          By : Tom Russell
//
//        Date : 14-Dec-03
//
// Description : Handles the set controls menu
//
//
//
////////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////////////////////
// Includes
////////////////////////////////////////////////////////////////////////////////
#include "setcontrolsmenu.hh"
#include "font.hh"

char *controlStrings[NUM_OF_CONTROLS] = { "Fire Weapon",
                                          "Change Weapon Up",
                                          "Change Weapon Down",
                                          "Use Jump Jets",
                                          "Use Shield",
                                          "Move Tank Left",
                                          "Move Tank Right",
                                          "Rotate Gun Left",
                                          "Rotate Gun Right",
                                          "Increase Gun Power",
                                          "Decrease Gun Power" };

// For the joystick Axis, certain command must be used in pairs, for example:
// if joystick left is set to 'move left' then joystick right is automatically
// set to 'move right'. The next table identifys which controls are linked
// together.

char linkedControls[NUM_OF_CONTROLS] = 
{
    -1,  // Fire
    2,   // Change Weapon Up
    1,   // Change Weapon Down
    -1,  // Jump jet
    -1,  // Shield
    6,   // Move left
    5,   // Move Right
    8,   // Rotate Left
    7,   // Rotate Right
    10,  // Increase Power
    9    // Decrease Power
};

// Names of all the special (non-ascii) keys on a normal keyboard. This is so
// we can display which one has been selected.

char *specialKeys[62] = 
{
    "Escape",
    "F1", "F2", "F3", "F4", "F5", "F6", "F7", "F8", "F9", "F10", "F11", "F12",
    "F13", "F14", "F15", "F16", "F17", "F18", "F19", "F20", "F21", "F22", "F23",
    "F24", "F25",
    "Up", "Down", "Left", "Right",
    "Left Shift", "Right Shift", "Left Control", "Right Control",
    "Left Alt", "Right Alt", "Tab", "Enter", "Backspace", "Insert", "Delete",
    "Page Up", "Page Down", "Home", "End", 
    "Keypad 0", "Keypad 1", "Keypad 2", "Keypad 3", "Keypad 4", "Keypad 5",
    "Keypad 6", "Keypad 7", "Keypad 8", "Keypad 9", "Keypad /", "Keypad *",
    "Keypad -", "Keypad +", "Keypad .", "Keypad =", "Keypad Enter"
};

// The (probable) names of the axes.
char *axes[8] = 
{
    "Joystick/Pad Right",
    "Joystick/Pad Left",
    "Joystick/Pad Up",
    "Joystick/Pad Down",
    "Axis 3 (-)",
    "Axis 3 (+)",
    "Axis 4 (-)",
    "Axis 4 (+)"
}; 

////////////////////////////////////////////////////////////////////////////////
// Public Member Functions
////////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////////////////////
//
// Function    : cSetControlsMenu
//
// Description : Constructor
//
////////////////////////////////////////////////////////////////////////////////
cSetControlsMenu::cSetControlsMenu
(
    cGame * game,
    int     layout
)
        : cMenu (game), _layout (layout)
{
    _controls = game->getControls ();

    // Create the buttons to change each control
    for (int i = 0; i < NUM_OF_CONTROLS; i++) 
    {
        _controlButtons[i] = new cTextButton (this, 
                                              -3.0f, 5.0f - i * 0.8f, 0.5f,
                                              controlStrings[i]);    

        _controlKey[i] = _controls->getControl (_layout, i);
    }

    _resetToDefaultsButton = new cTextButton (this, 
                                              0.0f, -5.0f, 0.7f, 
                                              "Reset To Defaults");

    _doneButton            = new cTextButton (this, 
                                              0.0f, -6.0f, 0.7f, 
                                              "Done");

    _waitingForKey = -1;
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : ~cSetControlsMenu
//
// Description : Destructor
//
////////////////////////////////////////////////////////////////////////////////
cSetControlsMenu::~cSetControlsMenu
(
)
{
    for (int i = 0; i < NUM_OF_CONTROLS; i++) 
    {
        _controls->setControl (_layout, i, _controlKey[i]);

        delete _controlButtons[i];
    }

    delete _doneButton;
    delete _resetToDefaultsButton;
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : update
//
// Description : Update the menu
//
////////////////////////////////////////////////////////////////////////////////
enumGameState
cSetControlsMenu::update
(
    double time
)
{
    updateBackground (time);


    for (int i = 0; i < NUM_OF_CONTROLS; i++)
    {
        if (_controlButtons[i]->update ())
        {
            // One of the change control buttons has been pressed. We now 
            // switch to 'waiting for input' mode. Disable all other buttons 
            // and listen for a key press/joystick input.
             
            // Mark which control we are changing
            _waitingForKey = i;

            for (int j = 0; j < NUM_OF_CONTROLS; j++) 
            {
                _controlButtons[j]->enable (false);
            }
        }
    }

    // Are we waiting for the user to specify a new key/axis for this control?
    if (_waitingForKey != -1)
    {    
        // If we are changing a keyboard layout, listen for the key press
        if (_layout < 2) 
        {
            // Get Keyboard button
            for (int i = 0; i < 318; i++)
            {
                if (GLFW_PRESS == _interface->getKey(i))
                {
                    _controlKey[_waitingForKey] = i;
                    _waitingForKey = -1;
                    
                    for (int j = 0; j < NUM_OF_CONTROLS; j++) 
                    {
                        _controlButtons[j]->enable (true);
                    }
                    break;
                }
            }
        }
        else
        {
            // We are changing a joystick layout so listen for a button/axis 
            // movement.
            for (int i = 2; i < _interface->numOfControllers (); i++)
            {
                int count;
                const unsigned char* buttons = glfwGetJoystickButtons(i - 2, &count);

                // Check the buttons
                for (int j = 0; j < count && j < 10; j++)
                {
                    if (buttons[j] == GLFW_PRESS)
                    {
                        if (_controlKey[_waitingForKey] >= 100 &&
                            linkedControls[_waitingForKey] != -1)
                        {
                            _controlKey[linkedControls[_waitingForKey]] = -1;    
                        }

                        _controlKey[_waitingForKey] = j;
                        _waitingForKey = -1;

                        for (int k = 0; k < NUM_OF_CONTROLS; k++) 
                        {
                            _controlButtons[k]->enable (true);
                        }
                        break;
                    }                    
                }

                if (_waitingForKey == -1)
                {
                    break;
                }
                
                // Check the axes
                int axisCount;
                const float* axesPtr = glfwGetJoystickAxes(i - 2, &axisCount);

                for (int j = 0; j < axisCount && j < 4; j++)
                {
                    // An axis is considered active if it is over half way in 
                    // a direction. This only applies to analogue axes. Digital
                    // axes can only be -1, 0, or 1
                    if (axesPtr[j] > 0.5)
                    {
                        // Axis controls are identified by numbers starting 
                        // from 100+
                        _controlKey[_waitingForKey] = 100 + (j * 2);
                        if (linkedControls[_waitingForKey] != -1)
                        {
                            _controlKey[linkedControls[_waitingForKey]] 
                                = 101 + (j * 2);    
                        }
                        _waitingForKey = -1;

                        // Re-enable all the other buttons
                        for (int k = 0; k < NUM_OF_CONTROLS; k++) 
                        {
                            _controlButtons[k]->enable (true);
                        }
                        break;
                    }
                    else if (axesPtr[j] < -0.5)
                    {
                        _controlKey[_waitingForKey] = 101 + (j * 2);
                        if (linkedControls[_waitingForKey] != -1)
                        {
                            _controlKey[linkedControls[_waitingForKey]] 
                                = 100 + (j * 2);
                        }
                        _waitingForKey = -1;

                        // Re-enable all the other buttons
                        for (int k = 0; k < NUM_OF_CONTROLS; k++) 
                        {
                            _controlButtons[k]->enable (true);
                        }
                        break;
                    }
                }

                if (_waitingForKey == -1)
                {
                    break;
                }
            }
        }
    }

    if (_doneButton->update ())
    {
        return (CONTROLLERS_MENU);
    }

    if (_resetToDefaultsButton->update ())
    {
        // Restore all keys to their defaults
        _controls->resetToDefault (_layout);

        for (int i = 0; i < NUM_OF_CONTROLS; i++) 
        {   
            _controlKey[i] = _controls->getControl (_layout, i);
        }
    }

    return (CURRENT_STATE);
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : draw
//
// Description : Draw the menu
//
////////////////////////////////////////////////////////////////////////////////
void
cSetControlsMenu::draw
(
)

{
    drawBackground ();

    glEnable (GL_BLEND);
  
    // Draw the dark boxes behind the text
    glColor4f (0.0f, 0.0f, 0.0f, 0.5f);

    glBegin (GL_QUADS);

    glVertex3f   (-7.0f, -4.0f, 0.0f);
    glVertex3f   ( 7.0f, -4.0f, 0.0f);
    glVertex3f   ( 7.0f,  6.0f, 0.0f);
    glVertex3f   (-7.0f,  6.0f, 0.0f);

    glVertex3f   (-7.0f, -6.6f, 0.0f);
    glVertex3f   ( 7.0f, -6.6f, 0.0f);
    glVertex3f   ( 7.0f, -4.4f, 0.0f);
    glVertex3f   (-7.0f, -4.4f, 0.0f);

    glEnd ();

    // Draw the dark brown boxes
    glColor4f (0.6f, 0.3f, 0.0f, 0.5f);

    for (int i = 0; i < NUM_OF_CONTROLS; i++)
    {
        glBegin (GL_QUADS);
        
        glVertex3f   (-6.0f, 5.3f - (0.8f * i), 0.0f);
        glVertex3f   ( 0.0f, 5.3f - (0.8f * i), 0.0f);
        glVertex3f   ( 0.0f, 4.6f - (0.8f * i), 0.0f);
        glVertex3f   (-6.0f, 4.6f - (0.8f * i), 0.0f);
        
        glVertex3f   ( 0.1f, 5.3f - (0.8f * i), 0.0f);
        glVertex3f   ( 6.0f, 5.3f - (0.8f * i), 0.0f);
        glVertex3f   ( 6.0f, 4.6f - (0.8f * i), 0.0f);
        glVertex3f   ( 0.1f, 4.6f - (0.8f * i), 0.0f);
        
        glEnd ();    
    } 

    glBegin (GL_QUADS);

    glVertex3f   (-4.0f, -5.4f, 0.0f);
    glVertex3f   ( 4.0f, -5.4f, 0.0f);
    glVertex3f   ( 4.0f, -4.6f, 0.0f);
    glVertex3f   (-4.0f, -4.6f, 0.0f);

    glVertex3f   (-4.0f, -6.4f, 0.0f);
    glVertex3f   ( 4.0f, -6.4f, 0.0f);
    glVertex3f   ( 4.0f, -5.6f, 0.0f);
    glVertex3f   (-4.0f, -5.6f, 0.0f);

    glEnd ();

    glDisable (GL_BLEND);
    
    _font->setSize (0.6f, 0.6f, 0.5f);
    _font->setColour (1.0f, 1.0f, 1.0f);

    if (_layout < 2) 
    {
        _font->printCentredAt (0.0f, 6.5f, "Editing Keyboard Layout %i", 
                               _layout + 1);
    }
    else
    {
        _font->printCentredAt (0.0f, 6.5f, "Editing Joystick Layout %i", 
                               _layout - 1);
    }

    if (_waitingForKey != -1) 
    {
        _font->setSize (0.6f, 0.6f, 0.5f);
        _font->setColour (0.5f, 0.5f, 0.5f);

        _font->printCentredAt (0.0f, -4.0f, 
                               "Press Button for '%s'", 
                               controlStrings[_waitingForKey]);
    }

    _font->setSize (0.5f, 0.5f, 0.4f);
    _font->setColour (0.0f, 1.0f, 1.0f);

    // Print the current controls
    for (int i = 0; i < NUM_OF_CONTROLS; i++)
    {
        if (_layout < 2)
        {
            // Keyboard
            if (_controlKey[i] > 256)
            {
                _font->printCentredAt (3.0f, 4.7f - i * 0.8f, "%s", 
                                       specialKeys[_controlKey[i] - 257]);
            }
            else if (_controlKey[i] > 32)
            {
                _font->printCentredAt (3.0f, 4.7f - i * 0.8f, "'%c'", 
                                       (char)_controlKey[i]);
            }
            else if (_controlKey[i] == 32)
            {
                _font->printCentredAt (3.0f, 4.7f - i * 0.8f, "Space");
            }
        }
        else
        {
            // Joystick
            if (_controlKey[i] == -1) 
            {
                _font->printCentredAt (3.0f, 4.7f - i * 0.8f, "<Undefined>");
            }
            else if (_controlKey[i] < 100)
            {
                // Button
                _font->printCentredAt (3.0f, 4.7f - i * 0.8f, "Joy Button %i", 
                                       _controlKey[i] + 1);
            }
            else
            {
                // Axis
                _font->printCentredAt (3.0f, 4.7f - i * 0.8f,
                                       axes[_controlKey[i] - 100]);
            }
            
        }
    }

    for (int i = 0; i < NUM_OF_CONTROLS; i++)
    {
        _controlButtons[i]->draw ();
    }

    _doneButton->draw ();
    _resetToDefaultsButton->draw ();
}
